File: C++/438. Find All Anagrams in a String.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        vector<int> vis1(26, 0);
        vector<int> vis2(26, 0);
        vector<int> ans;
        if(s.length() < p.length())
            return ans;
        for(int i=0;i<p.length();i++)
        {
            vis1[s[i]-'a']++;
            vis2[p[i]-'a']++;
        }
        if(vis1 == vis2)
            ans.push_back(0);
        for(int i=p.length();i<s.length();i++)
        {
            vis1[s[i]-'a']++;
            vis1[s[i-p.length()]-'a']--;
            if(vis1 == vis2)
                ans.push_back(i-p.length()+1);
        }
        return ans;
    }
};

File: Python/438. Find All Anagrams in a String.py
class Solution:
    def findAnagrams(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: List[int]
        """
        vis1 = [0]*26
        vis2 = [0]*26
        if len(p) > len(s):
            return []
        for i in range(len(p)):
            vis1[ord(p[i]) - ord('a')] += 1
            vis2[ord(s[i]) - ord('a')] += 1
        ans = list()
        if vis1 == vis2:
            ans.append(0)
        for i in range(len(p), len(s)):
            
            vis2[ord(s[i]) - ord('a')] += 1
            vis2[ord(s[i - len(p)]) - ord('a')] -= 1
            if vis1 == vis2:
                ans.append(i - len(p) + 1)
        return ans